### PR TITLE
feat(geyser): support dumping accounts to csv

### PR DESCRIPTION
**Pull request by 0xNineteen**
dump a snapshot loading to csv using geyser 

main command: `./zig-out/bin/geyser csv`

note: full dump of testnet snapshot is 75GB

- new `tryCollapse` method on recycle fba to collapse adjacent free records
- add docs on new usage
- add cli setup for geyser reader (`src/geyser/bin.zig`)
- add new EndOfSnapshotLoading enum to payload type
- add new geyser metrics to grafana

Screenshot 2024-10-03 at 10 27 42 AM

### CODE REVIEW

File: src/geyser/bin.zig [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+            // only dump accounts that match owner filters
+            if (maybe_owner_pubkeys) |owners| {
+                if (!owners.contains(account.owner)) {
+                    continue;
+                }
+            }
+            fmt_count += 120 + 5 * account.data.len;
+        }
+
+        var csv_string = try recycle_fba.allocator().alloc(u8, fmt_count);
```
Comment (InKryption) [inline src/geyser/bin.zig:201]:
```suggestion
        const csv_string = try recycle_fba.allocator().alloc(u8, fmt_count);
```

File: src/geyser/bin.zig [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    defer {
+        io_channel.deinit();
+        allocator.destroy(io_channel);
+    }
+
+    const io_handle = try std.Thread.spawn(.{}, csvDumpIOWriter, .{ exit, csv_file, io_channel, recycle_fba });
+    defer io_handle.join();
+    errdefer {
+        exit.store(true, .release);
+    }
```
Comment (InKryption) [inline src/geyser/bin.zig:168]: Any reason for the stylistic decision to write this as a block instead of as `errdefer exit.store(true, .release)`? It looks a bit glaring when it's right underneath `defer io_handle.join();`.

File: src/geyser/bin.zig [hunk 3]
```diff
@@ -0,0 +1,278 @@
+const std = @import("std");
+const sig = @import("sig");
+const cli = @import("zig-cli");
+
+pub const Config = struct {
+    pipe_path: []const u8 = sig.VALIDATOR_DIR ++ "geyser.pipe",
+    measure_rate_secs: u64 = 5,
+    geyser_bincode_buf_len: u64 = 1 << 29,
+    geyser_io_buf_len: u64 = 1 << 29,
+    owner_accounts: [][]const u8 = &.{},
```
Comment (InKryption) [inline src/geyser/bin.zig:10]: Does this have to be a mutable slice? Is it a limitation of `zig-cli`?

**Review by InKryption — CHANGES_REQUESTED** [review]
Some observations and questions

File: src/geyser/bin.zig [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        }
+
+        switch (std.meta.activeTag(payload)) {
+            .AccountPayloadV1 => {},
+            .EndOfSnapshotLoading => {
+                std.debug.print("recv end of snapshot loading signal\n", .{});
+                exit.store(true, .release);
+                break;
+            },
+        }
```
Comment (sonicfromnewyoke) [inline src/geyser/bin.zig:186]: it looks like the first 3 lines are extra.

just curious, why do you stop reading from Geyser at the end of the snapshot?

In my understanding, this event indicates switching the `is_startup` flag, on which relies the `update_account` function, from `true` to `false`.
I believe it's better to leave a comment to describe what should be implemented in the next step (e.g buffering and dumping of ongoing data)

File: src/geyser/readme.md [hunk 5]
```diff
@@ -7,6 +7,46 @@ opening a file-based pipe using the `mkfifo` syscall which is then
...
+
+```bash 
+zig build -Doptimize=ReleaseSafe
+./zig-out/bin/benchmark geyser
+```
+
+you can also benchmark an dummy reader
+
+```bash
+```bash
```
Comment (sonicfromnewyoke) [inline src/geyser/readme.md:27]: i guess, this is a typo
```suggestion
```bash
```

File: src/utils/allocators.zig [hunk 6]
```diff
@@ -11,7 +11,7 @@ pub fn RecycleFBA(config: struct {
         // (only used on init/deinit + arraylist expansion)
         backing_allocator: std.mem.Allocator,
         // this does the data allocations (data is returned from alloc)
-        alloc_allocator: std.heap.FixedBufferAllocator,
+        fba_allocator: std.heap.FixedBufferAllocator,
```
Comment (Rexicon226) [inline src/utils/allocators.zig:14]: The previous name was silly 😆 . You can name this one `fba`, it already has "allocator" in the name.

File: build.zig [hunk 7]
```diff
@@ -151,14 +151,15 @@ pub fn build(b: *Build) void {
 
     // geyser reader
     const geyser_reader_exe = b.addExecutable(.{
-        .name = "geyser_reader",
-        .root_source_file = b.path("src/geyser/reader.zig"),
+        .name = "geyser",
+        .root_source_file = b.path("src/geyser/bin.zig"),
```
Comment (Rexicon226) [inline build.zig:155]: Why not call it `main.zig`? It is the entry-point of the geyser reader.

**Review by Rexicon226 — CHANGES_REQUESTED** [review]
I skimmed it over, no comments on correctness.

File: src/geyser/bin.zig [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+            },
+        },
+    },
+} };
+
+pub fn main() !void {
+    try cli.run(&cli_app, std.heap.page_allocator);
+}
+
+pub fn getOwnerFilters(allocator: std.mem.Allocator) !?std.AutoArrayHashMap(sig.core.Pubkey, void) {
```
[the remark this answers is not in the document]
Comment (InKryption) [reply src/geyser/bin.zig:90]: I think returning a managed data structure can make sense, especially when we are juggling multiple allocators in certain contexts already.
Though my preference would actually be to accept the array list as an out parameter.

File: src/geyser/bin.zig [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+var measure_rate_option = cli.Option{
+    .long_name = "measure-rate",
+    .help = "rate at which to measure reads",
+    .value_ref = cli.mkRef(&config.measure_rate_secs),
+    .required = false,
+    .value_name = "measure_rate_secs",
+};
+
+const cli_app = cli.App{ .version = "0.0.19", .author = "Syndica & Contributors", .command = .{
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: 5b03899

File: src/geyser/bin.zig [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+    try cli.run(&cli_app, std.heap.page_allocator);
+}
+
+pub fn getOwnerFilters(allocator: std.mem.Allocator) !?std.AutoArrayHashMap(sig.core.Pubkey, void) {
+    const owner_accounts_str = config.owner_accounts;
+    if (owner_accounts_str.len == 0) {
+        return null;
+    }
+
+    var owner_pubkeys = std.AutoArrayHashMap(sig.core.Pubkey, void).init(allocator);
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig:107]: 5b03899

File: src/geyser/bin.zig [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+
+    // csv file to dump to
+    const dump_csv_path = try std.fmt.allocPrint(
+        allocator,
+        "{s}{s}",
+        .{ sig.VALIDATOR_DIR, "accounts.csv" },
+    );
+    defer allocator.free(dump_csv_path);
+
+    const csv_file = try std.fs.cwd().createFile(dump_csv_path, .{});
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: 5b03899

File: src/geyser/bin.zig (hunk 1 above)
Comment (0xNineteen) [reply src/geyser/bin.zig]: 5b03899

File: src/geyser/bin.zig (hunk 2 above)
Comment (0xNineteen) [reply src/geyser/bin.zig]: 5b03899

File: src/geyser/bin.zig (hunk 3 above)
Comment (0xNineteen) [reply src/geyser/bin.zig]: 5b03899

File: src/geyser/bin.zig [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+
+    return owner_pubkeys;
+}
+
+pub fn csvDump() !void {
+    const allocator = std.heap.c_allocator;
+
+    const metrics_thread = try sig.prometheus.spawnMetrics(allocator, 12355);
+    metrics_thread.detach();
+    std.debug.print("spawing metrics thread on port 12355\n", .{});
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: https://github.com/Syndica/sig/commit/5b03899d0b5773cedc6436dfc47597703ee9a196

File: src/geyser/bin.zig [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+    );
+    defer allocator.free(dump_csv_path);
+
+    const csv_file = try std.fs.cwd().createFile(dump_csv_path, .{});
+    defer csv_file.close();
+
+    std.debug.print("dumping to csv: {s}\n", .{dump_csv_path});
+
+    // setup reader
+    const exit = try allocator.create(std.atomic.Value(bool));
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: https://github.com/Syndica/sig/commit/5b03899d0b5773cedc6436dfc47597703ee9a196

File: src/geyser/bin.zig [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+    // preallocate memory for csv rows
+    const recycle_fba = try allocator.create(sig.utils.allocators.RecycleFBA(.{ .thread_safe = true }));
+    recycle_fba.* = try sig.utils.allocators.RecycleFBA(.{ .thread_safe = true }).init(allocator, 1 << 32);
+    defer {
+        recycle_fba.deinit();
+        allocator.destroy(recycle_fba);
+    }
+
+    // setup thread to write to csv
+    var io_channel = try sig.sync.Channel([]u8).create(allocator);
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: https://github.com/Syndica/sig/commit/5b03899d0b5773cedc6436dfc47597703ee9a196

File: src/geyser/bin.zig [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+    errdefer {
+        exit.store(true, .release);
+    }
+
+    // start to read from geyser
+    while (true) {
+        _, const payload = try reader.readPayload();
+        defer reader.resetMemory();
+
+        if (std.meta.activeTag(payload) == .EndOfSnapshotLoading) {
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: https://github.com/Syndica/sig/commit/5b03899d0b5773cedc6436dfc47597703ee9a196

File: src/geyser/bin.zig [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+            break;
+        }
+
+        switch (std.meta.activeTag(payload)) {
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: https://github.com/Syndica/sig/commit/5b03899d0b5773cedc6436dfc47597703ee9a196

File: src/geyser/bin.zig [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+            .AccountPayloadV1 => {},
+            .EndOfSnapshotLoading => {
+                std.debug.print("recv end of snapshot loading signal\n", .{});
+                exit.store(true, .release);
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: https://github.com/Syndica/sig/commit/5b03899d0b5773cedc6436dfc47597703ee9a196

File: src/geyser/bin.zig [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+    }
+}
+
+pub fn benchmark() !void {
+    const allocator = std.heap.c_allocator;
+
+    const pipe_path = config.pipe_path;
+    std.debug.print("using pipe path: {s}\n", .{pipe_path});
+
+    const exit = try allocator.create(std.atomic.Value(bool));
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: https://github.com/Syndica/sig/commit/5b03899d0b5773cedc6436dfc47597703ee9a196

File: src/utils/allocators.zig (hunk 6 above)
Comment (0xNineteen) [reply src/utils/allocators.zig:14]: even though the 'a' in 'fba' is allocator - i still prefer the explicit '_allocator' because its just so clear that its an allocator

File: src/geyser/bin.zig [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+                    &geyser_io_buf_len_option,
+                    &owner_accounts_option,
+                },
+            },
+        },
+    },
+} };
+
+pub fn main() !void {
+    try cli.run(&cli_app, std.heap.page_allocator);
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: [0775da7](https://github.com/Syndica/sig/pull/279/commits/0775da7ac8a41c3bb16610db0544818e3b1e5764)

File: src/accountsdb/db.zig [hunk 20]
```diff
@@ -453,7 +459,13 @@ pub const AccountsDB = struct {
         }
         self.logger.infof("total time: {s}", .{timer.read()});
 
-        self.logger.infof("combining thread accounts...", .{});
+        // if geyser, send end of data signal
+        if (self.geyser_writer) |geyser_writer| {
+            const end_of_snapshot = sig.geyser.core.VersionedAccountPayload{ .EndOfSnapshotLoading = {} };
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/accountsdb/db.zig]: [0775da7](https://github.com/Syndica/sig/pull/279/commits/0775da7ac8a41c3bb16610db0544818e3b1e5764)

File: src/geyser/readme.md (hunk 5 above)
Comment (0xNineteen) [reply src/geyser/readme.md]: [0775da7](https://github.com/Syndica/sig/pull/279/commits/0775da7ac8a41c3bb16610db0544818e3b1e5764)

File: src/geyser/bin.zig [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+    measure_rate_secs: u64 = 5,
+    geyser_bincode_buf_len: u64 = 1 << 29,
+    geyser_io_buf_len: u64 = 1 << 29,
+    owner_accounts: [][]const u8 = &.{},
+};
+
+var default_config = Config{};
+const config = &default_config;
+
+var owner_accounts_option = cli.Option{
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: nice idea - [0775da7](https://github.com/Syndica/sig/pull/279/commits/0775da7ac8a41c3bb16610db0544818e3b1e5764)

File: src/geyser/bin.zig (hunk 4 above)
Comment (0xNineteen) [reply src/geyser/bin.zig]: [0775da7](https://github.com/Syndica/sig/pull/279/commits/0775da7ac8a41c3bb16610db0544818e3b1e5764)

File: src/geyser/bin.zig (hunk 8 above)
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig:101]: imo if an extra allocator being passed around isnt gonna be critical (ie, in this case) then passing around a managed struct is ok - the api is way cleaner not having to worry about the allocator to pass to its methods (ie, `deinit`)

File: build.zig (hunk 7 above)
Comment (0xNineteen) [reply build.zig]: [1dffb78](https://github.com/Syndica/sig/pull/279/commits/1dffb78a0c3d8a3dceb66f4fa5e0de2bfcd41a29)

**Review by InKryption — APPROVED** [review]

**Review by sonicfromnewyoke — APPROVED** [review]
LGTM

File: src/utils/allocators.zig (hunk 6 above)
Comment (Rexicon226) [reply src/utils/allocators.zig:14]: fair enough, no problem from me

File: src/geyser/bin.zig (hunk 8 above)
[the remark this answers is not in the document]
Comment (Rexicon226) [reply src/geyser/bin.zig]: fair reasoning, it's just a nit.

File: src/geyser/bin.zig [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+    for (owner_accounts_str) |owner_str| {
+        const owner_pubkey = try sig.core.Pubkey.fromString(owner_str);
+        owner_pubkeys.putAssumeCapacity(owner_pubkey, {});
+    }
+
+    return owner_pubkeys;
+}
+
+pub fn csvDump() !void {
+    const allocator = std.heap.c_allocator;
```
[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: [d053fd6](https://github.com/Syndica/sig/pull/279/commits/d053fd6c8e497f8cbe719577768a08e9b88a4a8c)

File: src/geyser/bin.zig [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+
+    const metrics_thread = try sig.prometheus.spawnMetrics(allocator, 12355);
```
[the remark this answers is not in the document]
Comment (Rexicon226) [reply src/geyser/bin.zig]: can do later, resolving this for now.

File: src/geyser/bin.zig [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+    exit: *std.atomic.Value(bool),
+    csv_file: std.fs.File,
+    io_channel: *sig.sync.Channel([]u8),
+    recycle_fba: *sig.utils.allocators.RecycleFBA(.{ .thread_safe = true }),
+) !void {
+    const total_payloads_estimate: u64 = 405_721;
+    var payloads_written: u64 = 0;
+    var timer = try sig.time.Timer.start();
+
+    while (!exit.load(.acquire)) {
```
[the remark this answers is not in the document]
Comment (Rexicon226) [reply src/geyser/bin.zig]: could you do the errdefer as well, it's a bit confusing to a reader why one is releasing in an error state.

[the remark this answers is not in the document]
Comment (0xNineteen) [reply src/geyser/bin.zig]: [fce84e0](https://github.com/Syndica/sig/pull/279/commits/fce84e09048a9f500ace5fdca140899c2726c686)

**Comment by Rexicon226** [comment]
I'd like a chance to take a glance at the Grafana, but otherwise the code is good for me.

**Comment by Rexicon226** [comment]
I played around with the Grafana a bit, and I don't like the ugly overlapping values. It's also really unclear what those values mean.

What I personally think makes the most sense is to 
1. use `rate(` for the stalls
2. use a bar chart

this is what my end result looks like:
image
it's still really unclear to me what the values actually signify, so many we should add a label that explains the unit.

**Comment by 0xNineteen** [comment]
nice - changed to use rate + bar and added docs explaining the metrics 

probs want to document every metric we are recording 🤔

File: src/time/estimate.zig [hunk 25]
```diff
@@ -61,3 +61,61 @@ pub fn printTimeEstimate(
         });
     }
 }
+
+pub fn printTimeEstimateStd(
```
Comment (Rexicon226) [inline src/time/estimate.zig:65]: just a small nit, could we name this to be `printTimeEstimateStderr` instead.

**Review by Rexicon226 — APPROVED** [review]
LGTM, just one nit

File: src/time/estimate.zig (hunk 25 above)
Comment (0xNineteen) [reply src/time/estimate.zig]: [a535dd0](https://github.com/Syndica/sig/pull/279/commits/a535dd007aec80cee870207228b23aa83b8b4d1e)

**Review by Rexicon226 — APPROVED** [review]
Great!

**Review by InKryption — APPROVED** [review]
